modules/build/src/main/scala/scala/build/options/BuildOptions.scala

```diff
@@ -303,87 +370,108 @@ final case class BuildOptions(
+  ): Either[ScalaVersionError, String] =
+    maybeScalaVersionStringArg match {
+      case Some(scalaVersionStringArg) =>
+        val prefix =
+          if (Util.isFullScalaVersion(scalaVersionStringArg)) scalaVersionStringArg
+          else if (scalaVersionStringArg.endsWith(".")) scalaVersionStringArg
+          else scalaVersionStringArg + "."
+        val matchingStableVersions = versionPool.filter(_.startsWith(prefix)).map(Version(_))
+        if (matchingStableVersions.isEmpty)
+          Left(new InvalidBinaryScalaVersionError(
+            scalaVersionStringArg,
+            latestSupportedStableVersions
+          ))
+        else {
+          val validMaxVersions = maxSupportedStableScalaVersions
+            .filter(_.repr.startsWith(prefix))
+          val validMatchingVersions = {
+            val filtered = matchingStableVersions.filter(v => validMaxVersions.exists(v <= _))
+            if (filtered.isEmpty) matchingStableVersions
+            else filtered
+          }.filter(v => isSupportedVersion(v.repr))
+          if (validMatchingVersions.isEmpty)
+            Left(new UnsupportedScalaVersionError(
               scalaVersionStringArg,
               latestSupportedStableVersions
             ))
-          else {
-            val validMaxVersions = maxSupportedStableScalaVersions
-              .filter(_.repr.startsWith(prefix))
-            val validMatchingVersions = {
-              val filtered = matchingVersions.filter(v => validMaxVersions.exists(v <= _))
-              if (filtered.isEmpty) matchingVersions
-              else filtered
-            }.filter(v => isSupportedVersion(v.repr))
-            if (validMatchingVersions.isEmpty)
-              Left(new UnsupportedScalaVersionError(
-                scalaVersionStringArg,
-                latestSupportedStableVersions
-              ))
-            else
-              Right(validMatchingVersions.max.repr)
-          }
-        case None =>
-          val validVersions = allStableVersions
-            .map(Version(_))
-            .filter(v => maxSupportedStableScalaVersions.exists(v <= _))
-          if (validVersions.isEmpty)
-            Left(new NoValidScalaVersionFoundError(
-              allStableVersions,
+          else
+            Right(validMatchingVersions.max.repr)
+        }
+      case None =>
+        val validVersions = versionPool
+          .map(Version(_))
+          .filter(v => maxSupportedStableScalaVersions.exists(v <= _))
+        if (validVersions.isEmpty)
+          Left(new NoValidScalaVersionFoundError(
+            versionPool,
+            latestSupportedStableVersions
+          ))
+        else
+          Right(validVersions.max.repr)
+    }
+
+  private def matchNewestNonStableScalaVersion(
+    maybeScalaVersionStringArg: Option[String],
+    versionPool: Seq[String]
+  ): Either[ScalaVersionError, String] =
+    maybeScalaVersionStringArg match {
+      case Some(scalaVersionStringArg) =>
+        val matchingVersions =
```

Comment:
I think the `case Some(scalaVersionStringArg) =>` case should look like this (this should compile, and is simpler):
```scala
        val matches = versionPool.contains(scalaVersionStringArg)
        if (matches) {
          val isSupportedVersion0 = isSupportedVersion(scalaVersionStringArg)
          if (isSupportedVersion0)
            Right(scalaVersionStringArg)
          else
            Left(new UnsupportedScalaVersionError(
              scalaVersionStringArg,
              latestSupportedStableVersions
            ))
        }
        else
          Left(new InvalidBinaryScalaVersionError(
            scalaVersionStringArg,
            latestSupportedStableVersions
          ))
```

Reply:
Or even:

```scala
        val matches = versionPool.contains(scalaVersionStringArg)
        lazy val isSupportedVersion0 = isSupportedVersion(scalaVersionStringArg)
        if (matches && isSupportedVersion0)
          Right(scalaVersionStringArg)
        else if (matches)
          Left(new UnsupportedScalaVersionError(
            scalaVersionStringArg,
            latestSupportedStableVersions
          ))
        else
          Left(new InvalidBinaryScalaVersionError(
            scalaVersionStringArg,
            latestSupportedStableVersions
          ))
```

Reply:
Yes! Thank you very much!